README.md

kneesocks
=========

[![npm version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]

*Put some nice SOCKS on your HTTP only program*


# Motivation

Driven by my burning desire to tunnel `npm install` requests through a SOCKSv5 proxy (eg. [shadowsocks](http://shadowsocks.org/en/index.html)). Since `npm` only support HTTP proxy, I wrote this simple command line module to fire up a HTTP Proxy server, then use it to tunnel requests to my SOCKS proxy.


# Features

* Supports HTTP and HTTPS requests
* Built-in DNS cache
* 150 loc, does nothing else


# Install

`npm install kneesocks --production -g`


# Usage

`kneesocks port1 port2`, eg. `kneesocks 10001 10002`

* `port1` is the port your HTTP proxy server listen to on localhost.
* `port2` is the port of your existing SOCKS proxy, also on localhost.

Then you can set your npm config:

* `npm config set http http://127.0.0.1:10001`
* `npm config set https-proxy http://127.0.0.1:10001`

To revert proxy config, use `npm config delete http` and `npm config delete https-proxy`.

To debug, use `DEBUG=proxy kneesocks port1 port2`


# Limit

* Node v0.10+ required (Node v0.11+ preferred).
* Probably not the best solution for day-to-day browsing, due to lack of remote dns and websocket support.
* If you don't mind building binaries, use [DeleGate](http://www.delegate.org/delegate/) or [Polipo](http://www.pps.univ-paris-diderot.fr/~jch/software/polipo/) instead.


# License

MIT

[npm-image]: https://img.shields.io/npm/v/kneesocks.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/kneesocks
[travis-image]: https://img.shields.io/travis/bitinn/kneesocks.svg?style=flat-square
[travis-url]: https://travis-ci.org/bitinn/kneesocks


CHANGELOG.md

Changelog
=========

# 1.x release

## v1.0.4

- Fix: dependency and test cases

## v1.0.3

- Enhance: raise agent.maxSockets to 20

## v1.0.2

- Fix: package.json `bin` field
- Enhance: document

## v1.0.1

- Fix: should be node v0.11 requirement due to an existing bug :(

## v1.0.0

- Major: bump node requirement to v0.10


# 0.1.x release

## v0.1.0

- Major: initial public release


test/test.js

// test tools
var chai = require('chai');
var expect = chai.expect;

// dependencies
var http = require('http');
var net = require('net');

// fake socks proxy
var Socks = require('socksv5');

// test subject
var Kneesocks = require('../lib/kneesocks.js');
var proxy = new Kneesocks({ http: 10001, socks: 10002 });

describe('kneesocks', function() {

	before(function(done) {
		var server = Socks.createServer(function(info, accept, deny) {
			var socket = accept(true);

			if (info.dstPort == 80) {
				socket.write('HTTP/1.1 200 OK\r\n');
				socket.write('Connection: close\r\n');
				socket.write('Content-Type: text/plain\r\n');
				socket.write('\r\n');
			}

			socket.end();
		});
		server.useAuth(Socks.auth.None());
		server.listen(10002, '127.0.0.1', done);
	});

	describe('constructor', function() {
		it('should setup http proxy server', function() {
			expect(proxy.opts.http).to.equal(10001);
			expect(proxy.opts.socks).to.equal(10002);
			expect(proxy.agent).to.be.an.instanceof(Socks.HttpAgent);
		});
	});

	describe('start', function() {
		it('should start http proxy', function(done) {
			proxy.start(function() {
				expect(proxy.server).to.be.an.instanceof(http.Server);
				expect(proxy.running).to.be.true;
				done();
			});
		});
	});

	describe('stop', function() {
		it('should stop http proxy', function(done) {
			proxy.stop(function() {
				expect(proxy.running).to.be.false;
				done();
			});
		});
	});

	describe('proxy', function() {
		it('should accept http request', function(done) {
			proxy.start(function() {
				var req = http.request({
					host: '127.0.0.1'
					, port: 10001
					, method: 'GET'
					, path: 'http://localhost/'
				}, function(res) {
					expect(res.statusCode).to.equal(200);
					proxy.stop(function() {
						done();
					});
				});
				req.end();
			});
		});

		it('should accept CONNECT request', function(done) {
			proxy.start(function() {
				var req = http.request({
					host: '127.0.0.1'
					, port: 10001
					, method: 'CONNECT'
					, path: 'localhost:443'
				});

				req.on('connect', function(res, socket, head) {
					expect(res.statusCode).to.equal(200);
					proxy.stop(function() {
						done();
					});
				});

				req.end();
			});
		});
	});

});


lib/kneesocks.js

/**
 * kneesocks.js
 *
 * A http proxy that wraps around your socks4/5 proxy
 */

// wrap the global dns module
var dnscache = require('dnscache')({
	enable: true
	, ttl: 3600
	, cachesize: 5000
});

var Socks = require('socksv5');
var Agent = Socks.HttpAgent;
var HTTP = require('http');
var URL = require('url');
var debug = require('debug')('proxy');

module.exports = Kneesocks;

/**
 * Kneesocks server
 *
 * @param   Object  opts  Config
 * @return  Void
 */
function Kneesocks(opts) {

	this.opts = {
		http: opts.http || 10001
		, socks: opts.socks || 10002
	};

	// agent with socks support
	this.agent = new Agent({
		proxyHost: '127.0.0.1'
		, proxyPort: this.opts.socks
		, maxSockets: 20
		, auths: [ Socks.auth.None() ]
	});

}

/**
 * Start HTTP Proxy
 *
 * @param   Function  cb  Callback
 * @return  Void
 */
Kneesocks.prototype.start = function(cb) {

	if (this.running) {
		return;
	}

	var self = this;
	var opts = this.opts;

	// proxy http requests
	this.server = HTTP.createServer(function(req, res) {
		var options = URL.parse(req.url);

		var url = 'http://' + options.hostname + '/';
		debug('start: ' + url);

		// construct request using socks agent
		options = {
			hostname: options.hostname
			, port: options.port
			, path: options.path
			, headers: req.headers
			, method: req.method
			, agent: self.agent
		};

		debug('options:');
		debug(options);

		var request = HTTP.request(options, function(result) {
			debug('load: ' + url);
			res.writeHead(result.statusCode, result.headers);

			result.on('end', function() {
				debug('done: ' + url);
			});
			result.pipe(res);
		});

		request.on('error', function(err) {
			debug('error: ' + url);
			debug(err.stack);
		});

		req.pipe(request);
	});

	// proxy https requests
	this.server.on('connect', function(req, inputSocket, head) {
		// handle CONNECT, assume https
		var options = URL.parse('https://' + req.url);

		var url = 'https://' + options.hostname + '/';
		debug('start: ' + url);

		options = {
			host: options.hostname
			, port: options.port
			, proxyHost: '127.0.0.1'
			, proxyPort: opts.socks
			, auths: [ Socks.auth.None() ]
		};

		var outputSocket = Socks.connect(options);

		outputSocket.on('connect', function(socket) {
			inputSocket.write('HTTP/1.1 200 Connection established.\r\n');
			inputSocket.write('Proxy-Connection: close\r\n');
			inputSocket.write('Proxy-Agent: Kneesocks\r\n');
			inputSocket.write('\r\n');

			socket.write(head);
			debug('load: ' + url);

			socket.pipe(inputSocket);
			inputSocket.pipe(socket);
		});

		outputSocket.on('error', function(err) {
			debug('error: ' + url);
			debug(err.stack);
		});

		outputSocket.on('close', function() {
			debug('done: ' + url);
		});
	});

	// start proxy
	this.running = true;
	this.server.listen(opts.http, '127.0.0.1', cb);

	debug('http proxy listening on port: ' + opts.http);
	debug('connected socks proxy on port: ' + opts.socks);

};

/**
 * Stop HTTP Proxy
 *
 * @param   Function  cb  Callback
 * @return  Void
 */
Kneesocks.prototype.stop = function(cb) {

	if (this.running) {
		this.running = false;
		this.server.close(cb);
		debug('http proxy stopped, was on port: ' + this.opts.http);
	}

};
